larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
+        // ATTN in training mode, just return everything as a cosmic-ray
+        this->SelectAllPfos(crSliceHypotheses, selectedPfos);
+
+        unsigned int bestSliceIndex(std::numeric_limits<unsigned int>::max());
+        if (!this->GetBestSliceIndex(pAlgorithm, nuSliceHypotheses, crSliceHypotheses, bestSliceIndex)) return;
+
+        for (unsigned int sliceIndex = 0; sliceIndex < nSlices; ++sliceIndex)
+        {
+            const SliceFeatures &features(sliceFeaturesVector.at(sliceIndex));
+            if (!features.IsFeatureVectorAvailable()) continue;
+
+            LArSvmHelper::ProduceTrainingExample(m_trainingOutputFile, sliceIndex == bestSliceIndex, features.GetFeatureVector());
+        }
+
+        return;
+    }
+
+    this->SelectPfosByProbability(nuSliceHypotheses, crSliceHypotheses, sliceFeaturesVector, selectedPfos);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::GetSliceFeatures(const NeutrinoIdTool *const pTool, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, SliceFeaturesVector &sliceFeaturesVector) const
+{
+    for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
+        sliceFeaturesVector.push_back(SliceFeatures(nuSliceHypotheses.at(sliceIndex), crSliceHypotheses.at(sliceIndex), pTool));
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool NeutrinoIdTool::GetBestSliceIndex(const MasterAlgorithm *const pAlgorithm, const SliceHypotheses &nuSliceHypotheses, const SliceHypotheses &crSliceHypotheses, unsigned int &bestSliceIndex) const
+{
+    unsigned int nHitsInBestSlice(0);
+    unsigned int nNuHitsInBestSlice(0);
+    unsigned int nuNHitsTotal(0);
+   
     for (unsigned int sliceIndex = 0, nSlices = nuSliceHypotheses.size(); sliceIndex < nSlices; ++sliceIndex)
     {
-        const PfoList &sliceOutput((m_selectAllNeutrinos || (m_selectOnlyFirstSliceNeutrinos && (0 == sliceIndex))) ?
-            nuSliceHypotheses.at(sliceIndex) : crSliceHypotheses.at(sliceIndex));
+        if (nuSliceHypotheses.at(sliceIndex).size() != 1)
+            throw StatusCodeException(STATUS_CODE_OUT_OF_RANGE);
+
+        const PfoList &nuFinalStates(nuSliceHypotheses.at(sliceIndex).front()->GetDaughterPfoList());
 
-        selectedPfos.insert(selectedPfos.end(), sliceOutput.begin(), sliceOutput.end());
+        CaloHitList reconstructedHits;
+        this->Collect2DHits(nuFinalStates, reconstructedHits);
+        this->Collect2DHits(crSliceHypotheses.at(sliceIndex), reconstructedHits);
+
+        const unsigned int nNuHits(this->CountNeutrinoInducedHits(reconstructedHits));
+        nuNHitsTotal += nNuHits;
+        
+        // TODO may need to be careful with reproducibility here?
+        if (nNuHits > nNuHitsInBestSlice)
```

Comment:
Always good to look out for this. We should discuss if there's anything subtle going on. It looks like the input slices should always been in the same order, and then the input pfos in the same order... Is this right?

Otherwise need some sorting and/or tiebreakers.

Reply:
Yes I think you are right, provided the input slices are always given to the tool in the same order (which I believe they are) then this shouldn't be an issue. I just wanted to flag to to make sure!